// @azure/communication-signaling/src/constants.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

export const MAX_NUMBER_OF_TOKEN_FETCH_RETRIES = 3;

// GCC High gov cloud URLs
export const GCCHIGH_TROUTER_SERVICE_URL = "https://go.trouter.gov.teams.microsoft.us/v4/a";
export const GCCHIGH_REGISTRAR_SERVICE_URL =
  "https://registrar.gov.teams.microsoft.us/V3/registrations";

// DoD gov cloud URLs
export const DOD_TROUTER_SERVICE_URL = "https://go.trouter.dod.teams.microsoft.us/v4/a";
export const DOD_REGISTRAR_SERVICE_URL =
  "https://registrar.dod.teams.microsoft.us/V3/registrations";

// Int URLs
export const INT_TROUTER_SERVICE_URL = "https://go.trouter-int.skype.net/v4/a";
export const INT_REGISTRAR_SERVICE_URL =
  "https://edge.skype.net/registrar/testenv/v3/registrations";

// Public URLs
export const PUBLIC_TROUTER_SERVICE_URL = "https://go.trouter.skype.com/v4/a";
export const PUBLIC_REGISTRAR_SERVICE_URL =
  "https://edge.skype.com/registrar/prod/v3/registrations";

// Gov cloud types
export enum CloudType {
  Public = "Public",
  GccHigh = "GCC High",
  Dod = "DoD"
}

export enum CloudPrefix {
  OrgId = "orgid",
  Acs = "acs",
  Spool = "spool",
  GccHigh = "gcch",
  GccHighAcs = "gcch-acs",
  Dod = "dod",
  DodAcs = "dod-acs"
}
