// @azure/communication-react/dist/dist-esm/react-composites/src/composites/common/styles/PeoplePaneContent.styles.js
/**
 * @private
 */
export const peoplePaneContainerTokens = {
    childrenGap: '0.5rem'
};
/**
 * @private
 */
export const peoplePaneContainerStyle = {
    root: {
        position: 'relative',
        maxHeight: '100%',
        overflow: 'hidden'
    }
};
/**
 * @private
 */
export const participantListContainerStyles = {
    root: {
        overflowY: 'scroll'
    }
};
/**
 * @private
 */
export const copyLinkButtonContainerStyles = {
    root: {
        width: '100%',
        padding: '0.5rem 1rem'
    }
};
/**
 * @private
 */
export const copyLinkButtonStackStyles = {
    root: {
        marginLeft: '0.5rem',
        marginRight: '0.5rem'
    }
};
/**
 * @private
 */
export const copyLinkButtonStyles = {
    root: {
        height: '2.5rem',
        width: '100%'
    },
    textContainer: {
        display: 'contents'
    }
};
/**
 * @private
 */
export const linkIconStyles = {
    marginRight: '0.5rem'
};
/**
 * @private
 */
export const localAndRemotePIPStyles = {
    root: {
        marginRight: '1rem'
    }
};
/**
 * @private
 */
export const themedCopyLinkButtonStyles = (mobileView, theme) => ({
    root: {
        minHeight: mobileView ? '3rem' : '2.5rem',
        borderRadius: mobileView ? theme.effects.roundedCorner6 : theme.effects.roundedCorner4
    }
});
