// @fluentui/react-northstar/dist/commonjs/utils/positioner/isIntersectingModifier.js
"use strict";

exports.__esModule = true;
exports.isIntersectingModifier = void 0;

var _core = require("@popperjs/core");

var isIntersectingModifier = {
  name: 'is-intersecting-modifier',
  enabled: true,
  phase: 'main',
  requires: ['preventOverflow'],
  fn: function fn(_ref) {
    var state = _ref.state,
        name = _ref.name;
    var popperRect = state.rects.popper;
    var popperAltOverflow = (0, _core.detectOverflow)(state, {
      altBoundary: true
    });
    var isIntersectingTop = popperAltOverflow.top < popperRect.height && popperAltOverflow.top > 0;
    var isIntersectingBottom = popperAltOverflow.bottom < popperRect.height && popperAltOverflow.bottom > 0;
    var isIntersecting = isIntersectingTop || isIntersectingBottom;
    state.modifiersData[name] = {
      isIntersecting: isIntersecting
    };
    state.attributes.popper = Object.assign({}, state.attributes.popper, {
      'data-popper-is-intersecting': isIntersecting
    });
  }
};
exports.isIntersectingModifier = isIntersectingModifier;
