// @azure/communication-react/dist/dist-esm/react-components/src/components/VoiceOverButton.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { DefaultButton, mergeStyles, TooltipHost } from '@fluentui/react';
import React, { useState } from 'react';
import { iconWrapperStyle, buttonStyle } from './styles/VoiceOverButton.style';
import { useTheme } from '../theming';
import { isDarkThemed } from '../theming/themeUtils';
import { inputButtonTooltipStyle } from './styles/InputBoxComponent.style';
/**
 * @private
 */
export const VoiceOverButton = (props) => {
    const { onRenderIcon, onClick, ariaLabel, className, id, tooltipContent } = props;
    const [isHover, setIsHover] = useState(false);
    const mergedButtonStyle = mergeStyles(buttonStyle, className);
    const theme = useTheme();
    const calloutStyle = {
        root: {
            padding: 0
        },
        calloutMain: {
            padding: '0.5rem'
        }
    }; // Place callout with no gap between it and the button.
    const calloutProps = {
        gapSpace: 0,
        styles: calloutStyle,
        backgroundColor: isDarkThemed(theme) ? theme.palette.neutralLighter : ''
    };
    return React.createElement(TooltipHost, { hostClassName: inputButtonTooltipStyle, content: tooltipContent, calloutProps: Object.assign({}, calloutProps) },
        React.createElement(DefaultButton, { className: mergedButtonStyle, ariaLabel: ariaLabel, onClick: onClick, id: id, onMouseEnter: () => {
                setIsHover(true);
            }, onMouseLeave: () => {
                setIsHover(false);
            } }),
        React.createElement("div", { className: iconWrapperStyle }, onRenderIcon(isHover)));
};
